pkg/utils/vlantrunk.go

```diff
@@ -0,0 +1,127 @@
+package utils
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+/*
+vlan trunk support requires OOT i40e driver for Intel 700 series NICs.
+
+FROM - i40e README
+- trunk
+  Supports two operations: add and rem.
+  - add: adds one or more VLAN id into VF VLAN filtering.
+  - rem: removes VLAN ids from the VF VLAN filtering list.
+  Example 1: add multiple VLAN tags, VLANs 2,4,5,10-20, by PF, p1p2, on
+  a selected VF, 1, for filtering, with the sysfs support:
+  #echo add 2,4,5,10-20 > /sys/class/net/p1p2/device/sriov/1/trunk
+  Example 2: remove VLANs 5, 11-13 from PF p1p2 VF 1 with sysfs:
+  #echo rem 5,11-13 > /sys/class/net/p1p2/device/sriov/1/trunk
+  Note: for rem, if VLAN id is not on the VLAN filtering list, the
+  VLAN id will be ignored.
+
+REGEX vaildation string: ^[0-9]+([,\-][0-9]+)*$
+
+tested on following inputs:
+10
+3,5
+3,200
+10-30
+2,4,5
+2,4,5,10-20
+*/
+
+// IsValidTrunkInput checks if trunk input format valid or not
+func IsValidTrunkInput(inStr string) (bool, error) {
+
+	// Sanity check with REGEX validation
+	var validString = regexp.MustCompile(`^[0-9]+([,\-][0-9]+)*$`)
+	if !validString.MatchString(inStr) {
+		return false, fmt.Errorf("invalid input format")
+	}
+
+	// Extract vlan numbers and check range validation
+	entries := strings.Split(inStr, ",")
+	for i := 0; i < len(entries); i++ {
+		if strings.Contains(entries[i], "-") {
+			rng := strings.Split(entries[i], "-")
+			if len(rng) != 2 {
+				return false, fmt.Errorf("invalid range value %s", rng)
+			}
+			rngSt, err := strconv.Atoi(rng[0])
+			if err != nil {
+				return false, fmt.Errorf("failed to parse %s vlan id, start range is incorrect", rng[0])
+			}
+			rngEnd, err := strconv.Atoi(rng[1])
+			if err != nil {
+				return false, fmt.Errorf("failed to parse %s vlan id,, end range is incorrect", rng[1])
+			}
+
+			if !isValidVlan(rngSt) || !isValidVlan(rngEnd) || rngSt >= rngEnd {
+				return false, fmt.Errorf("invalid vlan range %d-%d", rngSt, rngEnd)
+			}
+		} else {
+			vlan, err := strconv.Atoi(entries[i])
+			if err != nil {
+				return false, fmt.Errorf("failed to parse vlan entry %s", entries[i])
+			}
+
+			if !isValidVlan(vlan) {
+				return false, fmt.Errorf("invalid vlan %d", vlan)
+			}
+
+		}
+	}
+
+	return true, nil
+}
+
+func isValidVlan(vlan int) bool {
+	return vlan > 0 && vlan < 4095
+}
+
+var trunkExecCommand = exec.Command
+var trunkFileFormat = "/sys/class/net/%s/device/sriov/%d/trunk"
+
+// AddVlanTrunk add vlan trunk for given VF(vfID) for a PF 'pfName' from value given in 'value' string
+func AddVlanTrunk(pfName string, vfID int, value string) error {
```

Comment:
Should the string representation of the VLANs to be allowed to be passed to the VF be decoupled from the method arguments ?

i.e have a `type TrunkRange struct {...}`

input format for the `trunk` file differs in Mellanox implementation which is only : 
`<add|rem> <start_vid> <end_vid>`

Reply:
We still need to accept user input from same json field in both format(Intel, mellanox). When and how the input validation will be done?

Reply:
My suggestion is to decouple user input format from the input format to the VLAN trunk methods.

i.e
- Have one format for the user (VLAN trunk user format) which allows him to specify VLAN trunk ranges (e,.g `5,7,14-23`)
- Have one format for the API (lets call it VLAN trunk API format) that reflects that but in a more "programatic" way, e.g a list of structs with a start and end range.
Each Vlan trunk implementation is free to convert this to whatever format is needed by its driver

user input validation and conversion to VLAN trunk API format can be done in `config.go:LoadConf()`

WDYT?